packages/experiments/src/components/Text/createComponent.tsx

```diff
@@ -0,0 +1,97 @@
+import { IClassNames, IStyleFunction, mergeStyleSets } from 'office-ui-fabric-react';
+import * as React from 'react';
+import { ITheme } from './theming/ITheme';
+import { ThemeConsumer } from './theming/ThemeProvider';
+
+export type IStyleFunction<TProps, TStyles> = (props: TProps) => Partial<TStyles>;
+
+export interface IPropsWithStyles<TProps, TStyles> {
+  styles?: IStyleFunction<TProps, TStyles> | Partial<TStyles>;
+}
+
+// Components should accept styling.
+export type IComponentProps<TProps, TStyles> = TProps & {
+  styles?: IStyleFunction<IPropsWithStyles<TProps, TStyles>, TStyles> | Partial<TStyles>;
+};
+
+export type IStyleProps<TProps, TStyles> = TProps & {
+  theme: ITheme;
+};
+
+// Views should accept processed classNames.
+export type IViewProps<TProps, TStyles> = TProps & {
+  classNames: IClassNames<TStyles>;
+};
+
+const augmentations: any = {};
+
+export interface IComponentOptions<TProps, TStyles, TStatics> {
+  displayName: string;
+  state?: React.ComponentType<IComponentProps<TProps, TStyles>>;
+  styles?: IStyleFunction<IStyleProps<TProps, TStyles>, TStyles> | Partial<TStyles>;
+  view?: React.ComponentType<IViewProps<TProps, TStyles>>;
+  statics?: TStatics;
+}
+
+function evaluateStyle<TProps, TStyles>(
+  props: TProps,
+  styles?: IStyleFunction<TProps, TStyles> | Partial<TStyles> | undefined
+): Partial<TStyles> | undefined {
+  if (typeof styles === 'function') {
+    return styles(props);
+  }
+
+  return styles;
+}
+
+// Helper function to tie them together.
+export function createComponent<TProps, TStyles, TStatics = {}>(
+  options: IComponentOptions<TProps, TStyles, TStatics>
+): React.StatelessComponent<IComponentProps<TProps, TStyles>> & TStatics {
+  const result: React.StatelessComponent<TProps> = (userProps: TProps) => {
+    const augmented = augmentations[options.displayName] || {};
+    const ComponentState = augmented.state || options.state;
+    const ComponentView = augmented.view || options.view;
+    const componentStyles = augmented.styles || options.styles;
+
+    ComponentView.displayName = ComponentView.displayName || options.displayName;
+
+    const content = (processedProps: IComponentProps<TProps, TStyles>) => {
+      const { styles } = processedProps;
+
+      return (
+        <ThemeConsumer>
```

Comment:
How does this work alongside the customizable decorator? My understand of customizable is that it injects the theme so that components use them. I'm curious if this approach replaces or complements that, and if it complements, how.

Reply:
This is an experiment in using the new React context. I'm not 100% clear on the path between the two - this is an experiment to decipher what that path looks like. Once the code is there and we can play with it, I think it will be more clear about what mods need to be made to customizer.

Reply:
Even with this, I think there is room for improvement. For example, having both `ThemeConsumer` and `ThemeContext.Consumer` layers ends up adding 2 extra react dom layers for every component... that's not ideal. Maybe there is a cleaner approach.